### PR TITLE
{compiler}[dummy] GCC 5.3.0-2.26 (REVIEW)

### PR DESCRIPTION
**Pull request by boegel**
dep for #2523

### CODE REVIEW

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 10 (10 easyconfigs in this PR)
Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/8266463bdcdd9bdd1b52 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 10 (10 easyconfigs in this PR)
Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/ee120aacaaad5906280c for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 10 (10 easyconfigs in this PR)
Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/b2543e71c4ab9e120ddb for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 10 (10 easyconfigs in this PR)
Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/13e4e5a0a7258cee3c41 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 10 (10 easyconfigs in this PR)
Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/5136fff21cbdcb86e220 for a full test report.

**Comment by boegel** [comment]
failing test reports are caused by download issues, do ignore...

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 7 out of 10 (10 easyconfigs in this PR)
Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/1df1ce5b76e0fce053cf for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 7 out of 10 (10 easyconfigs in this PR)
Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/168c50a2f89f3276098a for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 7 out of 10 (10 easyconfigs in this PR)
Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/bb9a42d3c28d39f45aa7 for a full test report.

**Comment by boegel** [comment]
Hmm, not sure what the issue is here...

```
/bin/sh ../libtool  --tag=CC   --mode=link gcc  -fPIC -O2 -march=native -version-info 2:0:0 -L/user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/lib64 -L/user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/lib -L/user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/Bison/3.0.4-GCCcore-5.3.0/lib -o libfl.la -rpath /user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/flex/2.6.0-GCCcore-5.3.0/lib libmain.lo libyywrap.lo  -lm -lm -lpthread
libtool: link: gcc -shared  -fPIC -DPIC  .libs/libmain.o .libs/libyywrap.o   -L/user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/lib64 -L/user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/lib -L/user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/Bison/3.0.4-GCCcore-5.3.0/lib -lm -lpthread  -O2 -march=native   -Wl,-soname -Wl,libfl.so.2 -o .libs/libfl.so.2.0.0
/bin/ld.gold: error: /user/scratchdelcatty/gent/gvo000/gvo00002/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.3.0/crtbeginS.o: unsupported reloc 42 against global symbol _ITM_deregisterTMCloneTable
/bin/ld.gold: error: /user/scratchdelcatty/gent/gvo000/gvo00002/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.3.0/crtbeginS.o: unsupported reloc 42 against global symbol _ITM_registerTMCloneTable
/bin/ld.gold: error: /user/scratchdelcatty/gent/gvo000/gvo00002/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.3.0/crtbeginS.o: unsupported reloc 42 against global symbol _Jv_RegisterClasses
/user/scratchdelcatty/gent/gvo000/gvo00002/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.3.0/crtbeginS.o(.text+0x1e): error: unsupported reloc 42
/user/scratchdelcatty/gent/gvo000/gvo00002/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.3.0/crtbeginS.o(.text+0x6b): error: unsupported reloc 42
/user/scratchdelcatty/gent/gvo000/gvo00002/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/GCCcore/5.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.3.0/crtbeginS.o(.text+0xeb): error: unsupported reloc 42
collect2: error: ld returned 1 exit status
make[2]: *** [libfl.la] Error 1
```

**Comment by boegel** [comment]
@ocaisa: does this ring any bells for you?

**Comment by ocaisa** [comment]
No, but I haven't tried binutils 2.26 and it looks like that's where the problem arises. I read the bug report at https://lists.debian.org/debian-glibc/2015/12/msg00248.html and it says that new relocations affect all static libs...is that a fundamental problem?

On 19 Feb 2016 22:49, "Kenneth Hoste" <notifications@github.com<mailto:notifications@github.com>> wrote:

@ocaisa<https://github.com/ocaisa>: does this ring any bells for you?

—
Reply to this email directly or view it on GitHub<https://github.com/hpcugent/easybuild-easyconfigs/pull/2524#issuecomment-186423937>.


------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------
Forschungszentrum Juelich GmbH
52425 Juelich
Sitz der Gesellschaft: Juelich
Eingetragen im Handelsregister des Amtsgerichts Dueren Nr. HR B 3498
Vorsitzender des Aufsichtsrats: MinDir Dr. Karl Eugen Huthmacher
Geschaeftsfuehrung: Prof. Dr.-Ing. Wolfgang Marquardt (Vorsitzender),
Karsten Beneke (stellv. Vorsitzender), Prof. Dr.-Ing. Harald Bolt,
Prof. Dr. Sebastian M. Schmidt
------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------

**Comment by boegel** [comment]
@ocaisa: the problem seems to be that `flex` being built with `GCC/5.3.0` didn't include a build dep on `binutils/2.26`; that's a hard requirement now, because of changes made in binutils 2.26: you need to build the whole stack on top of 2.26, you can't mix with the older 2.25...

going to retest this

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 10 out of 10 (10 easyconfigs in this PR)
Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/6a0f773191a9e4c478bc for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 10 out of 10 (10 easyconfigs in this PR)
Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/66fef9cd076a7301dd66 for a full test report.

**Comment by boegel** [comment]
@ocaisa: \o/

**Comment by boegel** [comment]
@wpoely86: please review?

File: easybuild/easyconfigs/b/binutils/binutils-2.26-GCCcore-5.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    ('Bison', '3.0.4'),
+    # zlib required, but being linked instatically, so not a runtime dep
+    ('zlib', '1.2.8'),
+    # use same binutils version that was used when building GCC toolchain, to 'bootstrap' this binutils
+    ('binutils', version, '', True)
+]
+
+# statically link with zlib, to avoid runtime dependency on zlib
+preconfigopts = 'LIBS="$EBROOTZLIB/lib/libz.a"'
+prebuildopts = 'LIBS="$EBROOTZLIB/lib/libz.a"'
```
Comment (wpoely86) [inline easybuild/easyconfigs/b/binutils/binutils-2.26-GCCcore-5.3.0.eb]: Check that the resulting binutils is statically linked

File: easybuild/easyconfigs/b/binutils/binutils-2.26.eb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    ('flex', '2.6.0'),
+    ('Bison', '3.0.4'),
+    # zlib required, but being linked instatically, so not a runtime dep
+    ('zlib', '1.2.8'),
+]
+
+# statically link with zlib, to avoid runtime dependency on zlib
+# further, add the system library path in the rpath: this should 'harden' the
+# resulting binutils to bootstrap GCC (no trouble when other libstdc++ is build etc)
+preconfigopts = 'LIBS="$EBROOTZLIB/lib/libz.a -Wl,-rpath=/lib64 -Wl,-rpath=/usr/lib64 -Wl,-rpath=/lib -Wl,-rpath=/usr/lib"'
```
Comment (wpoely86) [inline easybuild/easyconfigs/b/binutils/binutils-2.26.eb]: add `/usr/lib/x86_64-linux-gnu/` for Debian and friends.

Comment (wpoely86) [inline easybuild/easyconfigs/b/binutils/binutils-2.26.eb]: and `/lib` and `/lib64` can be dropped I think.

File: easybuild/easyconfigs/m/M4/M4-1.4.17-GCCcore-5.3.0.eb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  although it has some extensions (for example, handling more than 9 positional parameters to macros).
+ GNU M4 also has built-in functions for including files, running shell commands, doing arithmetic, etc."""
+
+toolchain = {'name': 'GCCcore', 'version': '5.3.0'}
+
+sources = [SOURCELOWER_TAR_GZ]
+source_urls = [GNU_SOURCE]
+
+# use same binutils version that was used when building GCC toolchain
+builddependencies = [('binutils', '2.25', '', True)]
```
Comment (ocaisa) [inline easybuild/easyconfigs/m/M4/M4-1.4.17-GCCcore-5.3.0.eb]: version `2.26`

File: easybuild/easyconfigs/g/GCCcore/GCCcore-5.3.0.eb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+homepage = 'http://gcc.gnu.org/'
+description = """The GNU Compiler Collection includes front ends for C, C++, Objective-C, Fortran, Java, and Ada,
+ as well as libraries for these languages (libstdc++, libgcj,...)."""
+
+toolchain = {'name': 'dummy', 'version': ''}
+
+mpfr_version = '3.1.3'
+
+source_urls = [
+    'http://ftpmirror.gnu.org/%(namelower)s/%(namelower)s-%(version)s',  # GCC auto-resolving HTTP mirror
```
Comment (ocaisa) [inline easybuild/easyconfigs/g/GCCcore/GCCcore-5.3.0.eb]: `namelower` is wrong, need to hard input `gcc` (like in the 4.9.3 version)

File: easybuild/easyconfigs/b/binutils/binutils-2.26-GCCcore-5.3.0.eb (hunk 1 above)
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.26-GCCcore-5.3.0.eb]: how? manually?

or via a sanity check command?

Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.26-GCCcore-5.3.0.eb]: Hah, it's not (to `libz`)...

```
$ ldd /user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/SL6/sandybridge/software/binutils/2.26-GCCcore-5.3.0/bin/ar
	linux-vdso.so.1 =>  (0x00007fff1d180000)
	/usr/$LIB/libsnoopy.so => /usr/lib64/libsnoopy.so (0x00002b331abeb000)
	libbfd-2.26.20160125.so => /user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/SL6/sandybridge/software/binutils/2.26-GCCcore-5.3.0/lib/libbfd-2.26.20160125.so (0x00002b331a9eb000)
	libz.so.1 => /lib64/libz.so.1 (0x00002b331adfa000)
	libfl.so.2 => /user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/SL6/sandybridge/software/flex/2.6.0-GCCcore-5.3.0/lib/libfl.so.2 (0x00002b331b010000)
	libm.so.6 => /lib64/libm.so.6 (0x00002b331b013000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00002b331b297000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00002b331b4b4000)
	libc.so.6 => /lib64/libc.so.6 (0x00002b331b6b9000)
	/lib64/ld-linux-x86-64.so.2 (0x00002b331a9c9000)
```

It is for binutils 2.25 w/ GCCcore 4.9.3 though:

```
 ldd /user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/SL6/sandybridge/software/binutils/2.25-GCCcore-4.9.3/bin/ar
	linux-vdso.so.1 =>  (0x00007ffc3afff000)
	/usr/$LIB/libsnoopy.so => /usr/lib64/libsnoopy.so (0x00002b09d3447000)
	libbfd-2.25.so => /user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/SL6/sandybridge/software/binutils/2.25-GCCcore-4.9.3/lib/libbfd-2.25.so (0x00002b09d3247000)
	libfl.so.2 => /user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/SL6/sandybridge/software/flex/2.5.39-GCCcore-4.9.3/lib/libfl.so.2 (0x00002b09d337e000)
	libm.so.6 => /lib64/libm.so.6 (0x00002b09d3656000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00002b09d38db000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00002b09d3af8000)
	libc.so.6 => /lib64/libc.so.6 (0x00002b09d3cfc000)
	/lib64/ld-linux-x86-64.so.2 (0x00002b09d3225000)
```

any ideas @wpoely86?

File: easybuild/easyconfigs/b/binutils/binutils-2.26.eb (hunk 2 above)
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.26.eb]: well, how do we know for sure?

we need a sanity check command to verify static linking to `libz`?

File: easybuild/easyconfigs/m/M4/M4-1.4.17-GCCcore-5.3.0.eb (hunk 3 above)
Comment (boegel) [inline easybuild/easyconfigs/m/M4/M4-1.4.17-GCCcore-5.3.0.eb]: oopd

Comment (boegel) [inline easybuild/easyconfigs/m/M4/M4-1.4.17-GCCcore-5.3.0.eb]: nice catch btw :)

File: easybuild/easyconfigs/g/GCCcore/GCCcore-5.3.0.eb (hunk 4 above)
Comment (boegel) [inline easybuild/easyconfigs/g/GCCcore/GCCcore-5.3.0.eb]: urgh... again, nice catch!

File: easybuild/easyconfigs/b/binutils/binutils-2.26.eb (hunk 2 above)
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.26.eb]: @wpoely86: ping?

Comment (wpoely86) [inline easybuild/easyconfigs/b/binutils/binutils-2.26.eb]: A sanity check for it would be nice. A simple `ldd bin/* | grep -c libz` would do.

Comment (wpoely86) [inline easybuild/easyconfigs/b/binutils/binutils-2.26.eb]: (for the static linking) The rpath is just fine.

Comment (wpoely86) [inline easybuild/easyconfigs/b/binutils/binutils-2.26.eb]: maybe easyblock?

Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.26.eb]: Yeah, it's probably a good idea to look into an easyblock for `binutils`, it's getting out of hand...

**Comment by boegel** [comment]
So, two options here: I rework this to (or open a new PR for) `GCC/5.3.0-2.25` (i.e. downgrade to the `binutils/2.25` we already have), and use that as a base for `intel/2016.02*` (i.e. #2523 and #2620), or I postpone this until after the EB v2.7.0 release (which implies not including easyconfigs for the latest Intel compilers in EB v2.7.0)...

Thoughts @wpoely86, @ocaisa?

**Comment by ocaisa** [comment]
What about just reworking #2620  as a 4.9.3-2.25 branch of the latest intel compilers and holding off on the 5.3.0-2.26 branch?

**Comment by boegel** [comment]
@ocaisa That's a good suggestion... I'll look into that.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 10 out of 10 (10 easyconfigs in this PR)
node2159.delcatty.os - Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/c9d7cc77943cde2aa62213f5057ba102 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 10 out of 10 (10 easyconfigs in this PR)
node2474.golett.os - Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/60e51e381911bdf402705c7d2d8e9e7a for a full test report.

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by boegel** [comment]
(finally) good to go now that the `binutils` easyblock is there, cfr https://github.com/hpcugent/easybuild-easyblocks/pull/907

Thanks for the review and feedback @wpoely86 and @ocaisa!

**Pull request merged**